File: cargo_bot_php/src/Service/Calculator.php
<?php

namespace CargoBot\Service;

use CargoBot\Database;
use CargoBot\Repository\SettingsRepo;

/**
 * Расчёт стоимости доставки.
 *
 * Ни одно число не задано в коде: тарифы, коэффициенты, сборы, скидки,
 * курсы валют и правила округления читаются из базы при каждом расчёте.
 * Поэтому правки из админ-панели применяются мгновенно.
 */
class Calculator
{
    private Database $db;
    private SettingsRepo $settings;

    public function __construct(Database $db)
    {
        $this->db = $db;
        $this->settings = new SettingsRepo($db);
    }

    /**
     * @return array<string,mixed>|null null, если подходящего тарифа нет
     */
    public function calculate(int $deliveryTypeId, int $categoryId, float $weight,
                              float $volume = 0.0, ?int $userId = null): ?array
    {
        $tariff = $this->findTariff($deliveryTypeId, $categoryId);
        if (!$tariff) {
            return null;
        }

        $deliveryType = $this->db->one('SELECT * FROM delivery_types WHERE id = :i', ['i' => $deliveryTypeId]);
        $category     = $this->db->one('SELECT * FROM categories WHERE id = :i', ['i' => $categoryId]);

        // 1. Объёмный вес
        $volCoef = $this->coefficient('volumetric', $deliveryTypeId, null);
        $volumetricWeight = ($volCoef !== null && $volume > 0)
            ? round($volume * $volCoef, 2)
            : 0.0;

        // 2. Оплачиваемый вес = max(фактический, объёмный, минимальный по тарифу)
        $chargeable = max($weight, $volumetricWeight, (float) $tariff['min_weight']);

        // 3. Базовая стоимость (с учётом коэффициента плотного груза)
        $price = (float) $tariff['price_per_kg'];
        $density = $this->coefficient('density', null, $categoryId);
        if ($density !== null) {
            $price = $price * $density;
        }
        $base = round($price * $chargeable, 2);

        // 4. Минимальная стоимость заказа
        $minCost = (float) $tariff['min_order_cost'];
        $minApplied = $base < $minCost;
        if ($minApplied) {
            $base = $minCost;
        }

        // 5. Дополнительные сборы, скидки, наценки
        $lines = [];
        $total = $base;
        $services = $this->db->all('SELECT * FROM extra_services WHERE is_active = 1 ORDER BY kind, id');
        foreach ($services as $s) {
            $sDt  = $s['delivery_type_id'] !== null ? (int) $s['delivery_type_id'] : null;
            $sCat = $s['category_id'] !== null ? (int) $s['category_id'] : null;
            if ($sDt !== null && $sDt !== $deliveryTypeId) {
                continue;
            }
            if ($sCat !== null && $sCat !== $categoryId) {
                continue;
            }
            $value = (float) $s['value'];
            if ($s['calc_type'] === 'percent') {
                $amount = round($base * $value / 100, 2);
            } elseif ($s['calc_type'] === 'per_kg') {
                $amount = round($value * $chargeable, 2);
            } else {
                $amount = $value;
            }
            if ($s['kind'] === 'discount') {
                $amount = -$amount;
            }
            $lines[] = ['name' => (string) $s['name'], 'amount' => $amount];
            $total += $amount;
        }
        $total = max($total, 0.0);

        // 6. Округление по правилам из настроек
        $mode = $this->settings->get('rounding_mode', 'none');
        $step = (float) ($this->settings->get('rounding_step', '1') ?: '1');
        $total = $this->applyRounding($total, (string) $mode, $step);

        // 7. Пересчёт в базовую валюту (TJS)
        $totalTjs = null;
        $rate = $this->rateToBase((string) $tariff['currency_code']);
        if ($rate !== null && $tariff['currency_code'] !== 'TJS') {
            $totalTjs = round($total * $rate, 2);
        }

        $result = [
            'tariff'            => $tariff,
            'delivery_type'     => $deliveryType,
            'category'          => $category,
            'weight'            => $weight,
            'volume'            => $volume,
            'volumetric_weight' => $volumetricWeight,
            'chargeable_weight' => $chargeable,
            'price_per_kg'      => round($price, 2),
            'base_cost'         => $base,
            'min_order_applied' => $minApplied,
            'lines'             => $lines,
            'total'             => $total,
            'currency_code'     => (string) $tariff['currency_code'],
            'total_tjs'         => $totalTjs,
            'delivery_time'     => $tariff['delivery_time'] ?? null,
        ];

        // 8. История расчётов
        $this->db->insert('calculations', [
            'user_id'           => $userId,
            'tariff_id'         => (int) $tariff['id'],
            'delivery_type'     => $deliveryType['name'] ?? null,
            'category'          => $category['name'] ?? null,
            'weight'            => $weight,
            'volume'            => $volume,
            'chargeable_weight' => $chargeable,
            'total'             => $total,
            'currency_code'     => (string) $tariff['currency_code'],
            'details'           => json_encode(['base' => $base, 'lines' => $lines], JSON_UNESCAPED_UNICODE),
            'created_at'        => $this->db->now(),
        ]);

        return $result;
    }

    /** Сначала точный тариф (тип + категория), затем общий (категория не указана). */
    private function findTariff(int $deliveryTypeId, int $categoryId): ?array
    {
        $exact = $this->db->one(
            'SELECT * FROM tariffs
             WHERE is_active = 1 AND delivery_type_id = :dt AND category_id = :cat
             ORDER BY id LIMIT 1',
            ['dt' => $deliveryTypeId, 'cat' => $categoryId]
        );
        if ($exact) {
            return $exact;
        }
        return $this->db->one(
            'SELECT * FROM tariffs
             WHERE is_active = 1 AND delivery_type_id = :dt AND category_id IS NULL
             ORDER BY id LIMIT 1',
            ['dt' => $deliveryTypeId]
        );
    }

    /**
     * Наиболее подходящий активный коэффициент:
     * сначала привязанный к типу доставки/категории, затем глобальный.
     */
    private function coefficient(string $code, ?int $deliveryTypeId, ?int $categoryId): ?float
    {
        $rows = $this->db->all(
            'SELECT * FROM coefficients WHERE is_active = 1 AND code = :c ORDER BY id',
            ['c' => $code]
        );
        foreach ($rows as $row) {
            $rDt  = $row['delivery_type_id'] !== null ? (int) $row['delivery_type_id'] : null;
            $rCat = $row['category_id'] !== null ? (int) $row['category_id'] : null;
            if (($rDt !== null && $rDt === $deliveryTypeId) || ($rCat !== null && $rCat === $categoryId)) {
                return (float) $row['value'];
            }
        }
        foreach ($rows as $row) {
            if ($row['delivery_type_id'] === null && $row['category_id'] === null) {
                return (float) $row['value'];
            }
        }
        return null;
    }

    /** Курс валюты к базовой (TJS). */
    private function rateToBase(string $currencyCode): ?float
    {
        $cur = $this->db->one('SELECT * FROM currencies WHERE code = :c', ['c' => $currencyCode]);
        if (!$cur) {
            return null;
        }
        if ((int) $cur['is_base'] === 1) {
            return 1.0;
        }
        $rate = $this->db->one(
            'SELECT rate FROM exchange_rates WHERE currency_id = :i ORDER BY updated_at DESC, id DESC LIMIT 1',
            ['i' => (int) $cur['id']]
        );
        return $rate ? (float) $rate['rate'] : null;
    }

    private function applyRounding(float $value, string $mode, float $step): float
    {
        if ($mode === 'none' || $step <= 0) {
            return round($value, 2);
        }
        $q = $value / $step;
        if ($mode === 'up') {
            $q = ceil($q);
        } elseif ($mode === 'down') {
            $q = floor($q);
        } else {
            $q = round($q);
        }
        return round($q * $step, 2);
    }
}

File: cargo_bot/bot/services/calculator.py
"""Сервис расчёта стоимости доставки.

Ни одно число не захардкожено: тарифы, коэффициенты, сборы, скидки,
курсы валют и правила округления читаются из базы данных при каждом расчёте,
поэтому изменения из админ-панели применяются мгновенно.
"""
import json
from dataclasses import dataclass, field
from decimal import ROUND_CEILING, ROUND_FLOOR, ROUND_HALF_UP, Decimal

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from bot.models import (
    Calculation, Category, Coefficient, Currency, DeliveryType, ExchangeRate,
    ExtraService, Tariff,
)
from bot.repositories import SettingsRepo


@dataclass
class CalcResult:
    tariff: Tariff
    delivery_type: DeliveryType
    category: Category
    weight: Decimal
    volume: Decimal
    volumetric_weight: Decimal
    chargeable_weight: Decimal
    price_per_kg: Decimal
    base_cost: Decimal
    min_order_applied: bool
    lines: list[tuple[str, Decimal]] = field(default_factory=list)  # сборы/скидки/наценки
    total: Decimal = Decimal(0)
    currency_code: str = "USD"
    total_tjs: Decimal | None = None
    delivery_time: str | None = None


def _dec(value) -> Decimal:
    return Decimal(str(value))


def _apply_rounding(value: Decimal, mode: str, step: Decimal) -> Decimal:
    if mode == "none" or step <= 0:
        return value.quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
    q = value / step
    rounding = {"up": ROUND_CEILING, "down": ROUND_FLOOR}.get(mode, ROUND_HALF_UP)
    return (q.quantize(Decimal("1"), rounding=rounding) * step).quantize(
        Decimal("0.01"), rounding=ROUND_HALF_UP
    )


class CalculatorService:
    def __init__(self, session: AsyncSession):
        self.session = session
        self.settings = SettingsRepo(session)

    async def _find_tariff(self, delivery_type_id: int, category_id: int) -> Tariff | None:
        """Сначала точный тариф (тип + категория), затем общий (тип, категория = NULL)."""
        stmt = select(Tariff).where(
            Tariff.is_active.is_(True),
            Tariff.delivery_type_id == delivery_type_id,
            Tariff.category_id == category_id,
        )
        tariff = await self.session.scalar(stmt)
        if tariff:
            return tariff
        stmt = select(Tariff).where(
            Tariff.is_active.is_(True),
            Tariff.delivery_type_id == delivery_type_id,
            Tariff.category_id.is_(None),
        )
        return await self.session.scalar(stmt)

    async def _coefficient(self, code: str, delivery_type_id: int | None = None,
                           category_id: int | None = None) -> Decimal | None:
        """Ищем наиболее специфичный активный коэффициент, затем глобальный."""
        stmt = select(Coefficient).where(
            Coefficient.is_active.is_(True), Coefficient.code == code
        )
        rows = list((await self.session.scalars(stmt)).all())

        for row in rows:  # специфичные (совпадение по типу доставки или категории)
            if (row.delivery_type_id and row.delivery_type_id == delivery_type_id) or \
               (row.category_id and row.category_id == category_id):
                return _dec(row.value)
        for row in rows:  # глобальные
            if row.delivery_type_id is None and row.category_id is None:
                return _dec(row.value)
        return None

    async def _rate_to_base(self, currency_code: str) -> Decimal | None:
        cur = await self.session.scalar(select(Currency).where(Currency.code == currency_code))
        if not cur:
            return None
        if cur.is_base:
            return Decimal(1)
        rate = await self.session.scalar(
            select(ExchangeRate).where(ExchangeRate.currency_id == cur.id)
            .order_by(ExchangeRate.updated_at.desc())
        )
        return _dec(rate.rate) if rate else None

    async def calculate(self, delivery_type_id: int, category_id: int,
                        weight: float, volume: float = 0,
                        user_id: int | None = None) -> CalcResult | None:
        tariff = await self._find_tariff(delivery_type_id, category_id)
        if not tariff:
            return None

        delivery_type = await self.session.get(DeliveryType, delivery_type_id)
        category = await self.session.get(Category, category_id)

        w, v = _dec(weight), _dec(volume)

        # 1. Объёмный вес
        vol_coef = await self._coefficient("volumetric", delivery_type_id=delivery_type_id)
        volumetric_weight = (v * vol_coef).quantize(Decimal("0.01")) if vol_coef and v > 0 else Decimal(0)

        # 2. Оплачиваемый вес = max(факт, объёмный, минимальный по тарифу)
        chargeable = max(w, volumetric_weight, _dec(tariff.min_weight or 0))

        # 3. Базовая стоимость (с учётом коэффициента плотного груза)
        price = _dec(tariff.price_per_kg)
        density = await self._coefficient("density", category_id=category_id)
        if density:
            price = price * density
        base = (price * chargeable).quantize(Decimal("0.01"))

        # 4. Минимальная стоимость заказа
        min_cost = _dec(tariff.min_order_cost or 0)
        min_applied = base < min_cost
        if min_applied:
            base = min_cost

        # 5. Доп. сборы, скидки, наценки
        stmt = select(ExtraService).where(ExtraService.is_active.is_(True))
        services = (await self.session.scalars(stmt)).all()
        lines: list[tuple[str, Decimal]] = []
        total = base
        for s in services:
            if s.delivery_type_id and s.delivery_type_id != delivery_type_id:
                continue
            if s.category_id and s.category_id != category_id:
                continue
            val = _dec(s.value)
            if s.calc_type == "percent":
                amount = (base * val / 100).quantize(Decimal("0.01"))
            elif s.calc_type == "per_kg":
                amount = (val * chargeable).quantize(Decimal("0.01"))
            else:
                amount = val
            if s.kind == "discount":
                amount = -amount
            lines.append((s.name, amount))
            total += amount
        total = max(total, Decimal(0))

        # 6. Округление по правилам из настроек
        mode = await self.settings.get("rounding_mode", "none")
        step = _dec(await self.settings.get("rounding_step", "1") or "1")
        total = _apply_rounding(total, mode or "none", step)

        # 7. Пересчёт в базовую валюту (TJS)
        total_tjs = None
        rate = await self._rate_to_base(tariff.currency_code)
        if rate and tariff.currency_code != "TJS":
            total_tjs = (total * rate).quantize(Decimal("0.01"))

        result = CalcResult(
            tariff=tariff, delivery_type=delivery_type, category=category,
            weight=w, volume=v, volumetric_weight=volumetric_weight,
            chargeable_weight=chargeable, price_per_kg=price.quantize(Decimal("0.01")),
            base_cost=base, min_order_applied=min_applied, lines=lines,
            total=total, currency_code=tariff.currency_code,
            total_tjs=total_tjs, delivery_time=tariff.delivery_time,
        )

        # 8. История расчётов
        self.session.add(Calculation(
            user_id=user_id, tariff_id=tariff.id,
            delivery_type=delivery_type.name if delivery_type else None,
            category=category.name if category else None,
            weight=float(w), volume=float(v),
            chargeable_weight=float(chargeable), total=float(total),
            currency_code=tariff.currency_code,
            details=json.dumps(
                {"lines": [(n, str(a)) for n, a in lines], "base": str(base)},
                ensure_ascii=False,
            ),
        ))
        await self.session.commit()
        return result
